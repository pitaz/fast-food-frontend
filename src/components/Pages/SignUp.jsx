import React, { Component } from "react";
import { connect } from "react-redux";
import { Link, Redirect } from "react-router-dom";
import { bindActionCreators } from "redux";
import PropTypes from "prop-types";
import { validateSignUpInput } from "../../validations/validations";
import signUp, { removeErrorMsg } from "../../actions/auth/signUp";

export class SignUp extends Component {
  state = {
    email: '',
    name: '',
    password: '',
    password_confirmation: '',
    errors: {}
  };

  onChange = event => {
    const { errors } = this.state;
    if (errors[event.target.name]) {
      const newErrors = Object.assign({}, errors);
      delete newErrors[event.target.name];
      this.setState({
        [event.target.name]: event.target.value,
        errors: newErrors
      });
    } else {
      this.setState({
        [event.target.name]: event.target.value
      });
    }
  };

  onSubmit = e => {
    const { action } = this.props;
    e.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {} });
      action.signupUser(this.state);
    }
  };

  isValid = () => {
    const { errors, isValid } = validateSignUpInput(this.state);
    if (!isValid) {
      this.setState({ errors, password: "", password_confirmation: "" });
    }

    return isValid;
  };

  render() {
    const { auth, error } = this.props;
    const { name, email, password, password_confirmation, errors } = this.state;

    if (auth) {
      return <Redirect to="/menu" />;
    }

    return (
      <div className="container">
        <div className="wrapper card">
    
          <form>
            <div className="card-header">
            {
              (error && error.message) ? (
                // <div className="error-cont" id="error">
                // <span id="closebtn">&times;</span> 
                  <p id="error-message">{error.message}</p>
                // </div>
              ) : ''
            }
              <p>Sign up</p>
            </div>
            <div className="card-content form-group">
              <label>
                <b>Username</b>
              </label>
              <input
                type="text"
                placeholder="Enter Username"
                name="name"
                field="name"
                value={name}
                onChange={this.onChange}
                required
              />
              <p className="error">{errors.name}</p>
              <label>
                <b>Email</b>
              </label>
              <input
                type="email"
                placeholder="Email"
                name="email"
                field="email"
                value={email}
                onChange={this.onChange}
                required
              />
              <p className="error">{errors.email}</p>
              <label>
                <b>Password</b>
              </label>
              <input
                type="password"
                placeholder="Enter Password"
                name="password"
                field="password"
                onChange={this.onChange}
                value={password}
                required
              />
              <p className="error">{errors.password}</p>
              <label>
                <b>Confirm Password</b>
              </label>
              <input
                type="password"
                placeholder="Confirm Password"
                name="password_confirmation"
                field="password_confirmation"
                onChange={this.onChange}
                value={password_confirmation}
                required
              />
              <p className="error">{errors.password_confirmation}</p>
              <button type="submit" onClick={this.onSubmit} >Submit </button>
              <h5>
                Already Registered? <Link to="/signin">Sign In</Link>
              </h5>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

SignUp.propTypes = {
  auth: PropTypes.bool,
  error: PropTypes.instanceOf(Object),
  action: PropTypes.instanceOf(Object)
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  error: state.auth.error
});

const matchDispatchToProps = dispatch => ({
  action: bindActionCreators(
    {
      signupUser: signUp,
      deleteError: removeErrorMsg
    },
    dispatch
  )
});

export default connect(
  mapStateToProps,
  matchDispatchToProps
)(SignUp);
