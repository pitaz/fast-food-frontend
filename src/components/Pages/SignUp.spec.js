/* eslint-disable react/no-children-prop */
/* eslint-disable no-undef */
import React from 'react';
import expect from 'expect';

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { SignUp } from './SignUp.jsx';


let wrapper;
let mockedState;
let props;

describe('The SignUp Component Test Suite', () => {
  beforeEach(() => {
  mockedState = {
    auth: {
      isAuthenticated: true,
      user: {
        id: 1,
        role: 'admin',
        email: 'peter@mail.com',
        iat: 1549688481,
        exp: 1549774881
      },
      error: {}
    },
  };
  props = {
    action: {
      deleteError: jest.fn(),
      signUpUser: jest.fn(),
    },
    error: {},
    auth: false,
  };
  });

  it('should match the snapshot', () => {
    wrapper = shallow(<SignUp {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render <SignUpForm /> Component', () => {
    wrapper = shallow(<SignUp {...props} />);
    expect(wrapper.length).toBe(1);
    expect(wrapper.find(SignUp)).toBeDefined();
  });


  it('should test for onchange', () => { 
    wrapper = shallow(<SignUp {...props} />);

    const event = {
      target: { name: 'email', value: 'test@test.com' },
      preventDefault: () => { },
    };

    wrapper.instance().onChange(event);
    expect(wrapper.state().email).toEqual('test@test.com');
    expect(wrapper.instance().state.email).toEqual(event.target.value);
  });

  
  it('should test for onsubmit', () => { 
    wrapper = shallow(<SignUp {...props} />);

    const event = {
      target: { name: 'email', value: 'testtest.com' },
      preventDefault: () => { },
    };

    wrapper.instance().onChange(event);
    expect(wrapper.state().error).toEqual(undefined);
    expect(wrapper.instance().state.email).toEqual(event.target.value);
  });

 it('should test the onSubmit function', () => {
    const event = { preventDefault: () => {} };
    wrapper = shallow(<SignUp {...props} />);
    const form = wrapper.find('form');
    form.simulate('submit', event);
    const loginUser = jest.fn(mockedState);
    const promise = new Promise((resolve) => {
      resolve(wrapper.instance().onSubmit);
    });
    promise.then(() => expect(loginUser).toHaveBeenCalledTimes(1));
  });
});
