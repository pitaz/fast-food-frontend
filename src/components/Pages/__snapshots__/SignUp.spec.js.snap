// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SignUp Component Test Suite should match the snapshot 1`] = `
<div
  className="container"
>
  <div
    className="wrapper card"
  >
    <form>
      <div
        className="card-header"
      >
        <p>
          Sign up
        </p>
      </div>
      <div
        className="card-content form-group"
      >
        <label>
          <b>
            Username
          </b>
        </label>
        <input
          field="name"
          name="name"
          onChange={[Function]}
          placeholder="Enter Username"
          required={true}
          type="text"
          value=""
        />
        <p
          className="error"
        />
        <label>
          <b>
            Email
          </b>
        </label>
        <input
          field="email"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          required={true}
          type="email"
          value=""
        />
        <p
          className="error"
        />
        <label>
          <b>
            Password
          </b>
        </label>
        <input
          field="password"
          name="password"
          onChange={[Function]}
          placeholder="Enter Password"
          required={true}
          type="password"
          value=""
        />
        <p
          className="error"
        />
        <label>
          <b>
            Confirm Password
          </b>
        </label>
        <input
          field="password_confirmation"
          name="password_confirmation"
          onChange={[Function]}
          placeholder="Confirm Password"
          required={true}
          type="password"
          value=""
        />
        <p
          className="error"
        />
        <button
          onClick={[Function]}
          type="submit"
        >
          Submit 
        </button>
        <h5>
          Already Registered? 
          <Link
            replace={false}
            to="/signin"
          >
            Sign In
          </Link>
        </h5>
      </div>
    </form>
  </div>
</div>
`;
